Implement an amp-mega-menu component

leafsy opened the pull request:
I2I: #24814 

Implementation of `<amp-mega-menu>` component under the experiment name `amp-mega-menu`, with support for rendering dynamic data using `amp-list` and templating.

Design doc with API: https://docs.google.com/document/d/1z2Bq8fia7vAUWn8Enpbj0beX-DqUC-mzTRnuX3rjLjk/edit?usp=sharing

Non-responsive demo:
https://amp-mega-menu.firebaseapp.com/
Responsive demo (via media query with `amp-sidebar`):
https://amp-responsive-menu.firebaseapp.com/
Example file will be added soon.

/cc @cathyxz @sparhami @wassgha

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+  createScreenReaderCloseButton() {
+    const ariaLabel =
+      this.element.getAttribute('data-close-button-aria-label') ||
+      'Close the mega menu';
```
Pardon my lack of familiarity with some `amphtml` conventions, but do we have any approach to internationalization for text strings like this (even though there's a way to set it via attribute)?

Irrespective of that, maybe this would make sense as a constant.

leafsy replied:
afaik there's no unified approach to internationalizing these strings; I can make this a constant.

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+
+const TAG = 'amp-mega-menu';
+
+export class AmpMegaMenu extends AMP.BaseElement {
```
Nit: class description comment would be good here, possibly with code example

See the [amp-stream-gallery](https://github.com/ampproject/amphtml/pull/24388/files#diff-35ed6523926d6b1a5d14031ad92de542R60) component for a good example

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+    this.action_ = null;
+
+    /** @private {number|string} */
+    this.prefix_ = element.id ? element.id : Math.floor(Math.random() * 100);
```
Nit: `element.id || Math.floor(Math.random() * 100)`

Also, I'm uncertain if we have existing methods/standard for generating a UUID that could be used here.

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FIXED_HEIGHT;
```
I believe `===` is preferred

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+    this.documentElement_.addEventListener('keydown', event => {
+      // Collapse mega menu on ESC.
+      if (event.key == Keys.ESCAPE) {
+        if (this.collapse_()) {
```
Nit: `if (event.key === Keys.ESCAPE && this.collapse_()) ...`

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+    const mask = this.document_.createElement('div');
+    mask.classList.add('i-amphtml-mega-menu-mask');
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
```
Very helpful comments, nice! :+1:

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.css:
```diff
@@ -0,0 +1,74 @@
+ * limitations under the License.
+ */
+
+header {
```
Do we really want to apply these styles to all `<header>` elements, or only those within the mega-menu?

leafsy replied:
Good point; per suggestions from design review, I'll add a css class to the mask parent to ensure z-index styling, and avoid using the header tag as a CSS selector.

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.css:
```diff
@@ -0,0 +1,74 @@
+/**
```
Does everything in this file need to be `!important`? (Genuinely asking, not intimately familiar with CSS conventions in `amphtml`

leafsy replied:
I think the process is usually start off by making the properties non-overridable, then release any that has a need for user customization.

rcebulko commented on extensions/amp-mega-menu/0.1/amp-mega-menu.css:
```diff
@@ -0,0 +1,74 @@
+
+amp-mega-menu {
+  z-index: 1000 !important;
+  white-space: nowrap !important;
```
Does `amp-mega-menu` inherit a `position` from somewhere else? If `amp-mega-menu` is `position: static`, I don't think `z-index` applies?

leafsy replied:
yep, it inherits `position: relative` from `ampshared.css`.

rcebulko commented:
Also, since you're creating a new extension directory, please make yourself (and probably

cathyxz commented on extensions/amp-mega-menu/0.1/test/test-amp-mega-menu.js:
```diff
@@ -0,0 +1,177 @@
+
+      for (let i = 0; i < items.length; i++) {
+        const li = doc.createElement('li');
+        li.innerHTML = items[i];
```
We can use `htmlFor` instead of directly setting `innerHTML`. =)

leafsy replied:
it turns out `htmlFor` cannot take in a template with dynamic `${expression}`, so I removed the for loop and just wrote out the items; this seems more readable anyway imo.

cathyxz commented on extensions/amp-mega-menu/0.1/test/test-amp-mega-menu.js:
```diff
@@ -0,0 +1,177 @@
+      element = getAmpMegaMenu();
+      doc.body.appendChild(element);
+
+      toggleExperiment(win, 'amp-mega-menu', true);
```
Let's add a TODO here to remove the experiment toggle code post launch. I fact, we should add a cleanup issue for each experiment we introduce and reference the issue number for cleanup actions in the code.

cathyxz commented on extensions/amp-mega-menu/0.1/test/test-amp-mega-menu.js:
```diff
@@ -0,0 +1,177 @@
+      expect(heading.getAttribute('aria-expanded')).to.equal('false');
+    });
+
+    it('should be navigable by left/right arrow keys when a heading has focus', () => {
```
We should use `async` `await` here to wait for the promise to return. E.g.: 

```
async () => {
...
await promise; 
} 
```

cathyxz replied:
Should the other tests in this unit test suite also be waiting for build callback to complete?

rcebulko replied:
Adding onto cathyxz, I believe tests that run asynchronous callbacks run the risk of executing and, if the promise fails, never calling the `expect`s. IIRC there are two ways around this, either with a callback function ie.

```javascript
it('..', async done => {
  ...
  done();
}
```

or by returning a promise, in which case Mocha waits for the promise to resolve.

cathyxz commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+   * @param {!Event} event click event.
+   */
+  handleHeadingClick_(event) {
+    event.preventDefault();
```
Doesn't have to be in this PR, but we should define and document behavior for elements inside headings that have tap actions registered.

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.css:
```diff
@@ -0,0 +1,74 @@
+/**
```
Please run `gulp get-zindex` in order to update `css/Z_INDEX.md`

wassgha commented on tools/experiments/experiments.js:
```diff
@@ -119,6 +119,10 @@ const EXPERIMENTS = [
   },
+  {
+    id: 'amp-mega-menu',
+    name: 'AMP extension for a mega menu with expandable items',
```
Please link to your cleanup issue here

cathyxz commented:
To @wassgha 's point, @leafsy let's add your demo to the `tests/manual/` file in this PR so we can make use of the PR deploy bot.

leafsy commented:
Really appreciate everyone's feedback!
@cathyxz I've added the manual test file and moved logic from buildCallback to layoutCallback;
@wassgha indeed I'll add documentation and validation (+testing) in a separate PR.

leafsy commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,322 @@
+      setModalAsClosed(content);
+      content.setAttribute('aria-modal', 'false');
+    });
+    item.removeAttribute('open');
```
I believe this is ok since the attribute gives users a way to style the menu heading based on whether its content is expanded.

wassgha commented on test/manual/amp-mega-menu.amp.html:
```diff
@@ -0,0 +1,138 @@
+    <div id="title">amp-mega-menu</div>
+
+    <amp-mega-menu layout="fixed-height" height="55">
+      <amp-list layout="fixed-height" height="600" src="/list/ecommerce-nested-menu" items="." single-item>
```
Having the `amp-list` be of `fixed-height` means that elements under it won't be clickable at all (whether the mask is open or not). You can see on your demo that the `Visit AMP` link under the menu is not clickable because of this. We need to find a way around this (either toggle `visibility` on this element (not good since `amp-list` is not part of `amp-mega-menu`'s expected hierarchy), have the `amp-mega-menu` be `overflow: hidden` unless open, or something else.

leafsy replied:
That's a great observation (and a major oversight on my part). Keeping `overflow: hidden` unless the menu is open seems to be a suitable solution.

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,363 @@
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
```
For reference, `amp-sidebar` appends it to the document's body

leafsy replied:
Without header the mask will be appended to the mega menu element itself; this would still keep all elements inside the mega menu above the mask (though user might need to set background color on the menu items for it to look right). I believe with #24827 we want to move away from appending things directly under body, due to concerns from amp-story (even though in this case mega menu will probably not be used within stories).

wassgha replied:
Could you please test and see if the behavior works correctly without a header element? It seems to me like it wouldn't given that the mega menu is position relative and the mask inside it will be absolute but relative to it

leafsy replied:
I've made a [quick recording](https://vimeo.com/364689577) of what it looks like without a header. The mask has position fixed, so it should still cover the entire page (if that's the issue).

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,363 @@
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
+    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+    maskParent.appendChild(mask);
```
How does this interact with other masks?

leafsy replied:
The mask will be below masks for components like sidebar or lightbox gallery. Are there specific concerns you have with the interaction?

cathyxz replied:
Here are a couple of interesting use cases to test (or ban via validator): 
1. `amp-lightbox-gallery` triggered by a lightbox-ed image within `amp-mega-menu`. 
2. `amp-sidebar` opened by a click action while `amp-mega-menu` is open. 
3. `amp-lightbox` triggered by clicking on something inside `amp-mega-menu`. I can actually imagine people wanting to do this for popup offers, though I'm unclear on whether we want to encourage this kind of UX. 
4. Multiple `amp-mega-menu`s done for purpose of building multiline menus. 
5. `amp-mega-menu` inside `amp-lightbox`. If this is not a valid use case, we should ban it via validator. 

We should (feel free to do so in a follow up PR): 
1. Verify these cases don't break or result in awkward / bad-looking UX. 
2. If they do, either fix or restrict them.

leafsy replied:
Good call! I'll go through these cases and include relevant manual tests as part of validation pr.

leafsy commented on test/manual/amp-mega-menu.amp.html:
```diff
@@ -0,0 +1,138 @@
+    </amp-mega-menu>
+  </header>
+
+  <amp-img src="https://cdn.pixabay.com/photo/2015/12/15/09/04/banner-1093909__340.jpg" width="100" height="20" layout="responsive"></amp-img>
```
Good point! I'll update links for all images to use the sources you mentioned below.

leafsy commented on test/manual/amp-mega-menu.amp.html:
```diff
@@ -0,0 +1,138 @@
+    <div id="title">amp-mega-menu</div>
+
+    <amp-mega-menu layout="fixed-height" height="55">
+      <amp-list layout="fixed-height" height="600" src="/list/ecommerce-nested-menu" items="." single-item>
```
It does feel a bit out of place, though I'm leaning towards keeping the json here for now since it will be used by multiple manual tests and possibly examples. @cathyxz

cathyxz replied:
There are two options. One is to keep it in the routes file like you currently have. The other is to add a static `.json` file to a local folder (e.g. you can create a folder within `test/manual` for `amp-mega-menu`). For example within the set of amp-list examples, we have `test/manual/amp-list/amp-list-data.json`. 

I think that so long is the JSON blob isn't ridiculously large, this is fine.

cathyxz commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,363 @@
+const TAG = 'amp-mega-menu';
+
+/** @const {string} */
+const ARIA_LABEL_CLOSE = 'Close the mega menu';
```
Even though we call it a mega menu, I don't think screenreader users think of it as a mega menu. I would just say "close the menu". In fact, I would consider going one step further and say "Close the ${title} menu".

leafsy replied:
Is `${title}` in this case the menu item headings (e.g. Clothing, Movies, etc.)? If so I'm not sure of the best way to get those values. Perhaps by using `heading.textContent`, but then we'd need to enforce the title text to be directly in the menu button and not one of its children.

cathyxz replied:
True. Hmm... tricky. In that case, let's just remove the `mega`, since `mega-menu` is more of a developer construct and not necessarily end user terminology and we can think about adding an option to customise the aria label if there's a strong need for it later.

rcebulko commented on extensions/amp-mega-menu/OWNERS.yaml:
```diff
@@ -0,0 +1,5 @@
+# https://github.com/ampproject/amp-github-apps/blob/master/owners/OWNERS.example.yaml
+
+- leafsy
+- ampproject/wg-ui-and-a11y
```
Update: The OWNERS bot has been migrate to [this JSON format]. Please replace this file with `extensions/amp-mega-menu/OWNERS` containing:

```javascript
// For an explanation of the OWNERS rules and syntax, see:
// https://github.com/ampproject/amp-github-apps/blob/master/owners/OWNERS.example

{
  rules: [
    { name: 'leafsy' },
    { name: 'ampproject/wg-ui-and-a11y' }
  ]
}
```

kristoferbaxter commented on build-system/server/routes/list.js:
```diff
@@ -180,7 +180,7 @@ router.get('/ecommerce-nested-menu', function(req, res) {
       {
         'title': 'Clothing, Shoes, Jewelry \u0026 Watches',
-        'image': 'https://cdn.orvis.com/images/082119_w_new.jpg',
+        'image': 'https://picsum.photos/id/535/367/267',
```
In a future PR can we make these served by the same system providing the JSON?

cathyxz replied:
We have some sample images in `example/img`. We haven't been super consistent about downloading all of the images we use into that folder and serving from that folder. We often use picsum simply because they have high quality pretty photos. It could be worthwhile to download a set of UI-friendly jpgs, organize, and document them for testing purposes.

cathyxz replied:
Filed https://github.com/ampproject/amphtml/issues/25045 as a follow up to address in a future PR.

kristoferbaxter commented on extensions/amp-mega-menu/0.1/amp-mega-menu.css:
```diff
@@ -0,0 +1,74 @@
+ */
+
+amp-mega-menu {
+  z-index: 1000 !important;
```
In the future, can we make these values (1000), based on a system with variables? This would help authors to understand what the semantic meaning of `1000` when used as a z-index.

kristoferbaxter commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,363 @@
+    this.maskElement_ = mask;
+
+    this.documentElement_.addEventListener('click', () => this.collapse_());
+    this.documentElement_.addEventListener('keydown', event => {
```
Do these events get cleaned up anywhere besides page deconstruction?

cathyxz replied:
Good catch! We should unregister them in `unlayoutCallback`.

leafsy replied:
filed #25047 to make the change and verify via unit test.

kristoferbaxter commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,363 @@
+    });
+    item.setAttribute('open', '');
+    this.element.setAttribute('open', '');
+    this.maskElement_.setAttribute('open', '');
```
Nit (and not required to be changed!): Can be expressed as a [].forEach since the operation is the same per element.

`[item, this.element, this.maskElement_].forEach(e=>e.setAttribute('open', ''));`

kristoferbaxter commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,363 @@
+    });
+    item.removeAttribute('open');
+    this.element.removeAttribute('open');
+    this.maskElement_.removeAttribute('open');
```
Nit: Similar to the adding of the `open` attribute, this could be represented as an array since the operation is the same per element.

cathyxz commented:
Friendly ping @rcebulko for review

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,359 @@
+   */
+  registerMenuItem_(item, heading, content) {
+    item.classList.add('i-amphtml-mega-menu-item');
+    this.itemCount_++;
```
Is there a case where this is different from `this.items_.length`?

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,359 @@
+   */
+  registerMenuItem_(item, heading, content) {
+    item.classList.add('i-amphtml-mega-menu-item');
+    this.itemCount_++;
```
Is there a case where this is different from `this.items_.length`?

leafsy replied:
One theoretical case is when dom update removes some items and then adds different items, though not sure if it's actually possible.

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,359 @@
+    this.items_ = toArray(scopedQuerySelectorAll(this.element, 'nav > * > li'));
+    this.items_.forEach(item => {
+      // skip if the item has already been registered.
+      if (item.classList.contains('i-amphtml-mega-menu-item')) {
```
I'd use `.filter` here

wassgha commented on extensions/amp-mega-menu/0.1/amp-mega-menu.js:
```diff
@@ -0,0 +1,359 @@
+        this.action_.hasAction(heading, 'tap', item)
+      ) {
+        return;
+      }
```
Do we still want the item in `this.items_` if it's not really registered? If not, we can get rid of `this.itemCount_`
